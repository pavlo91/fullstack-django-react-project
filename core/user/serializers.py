from rest_framework import serializers

from core.abstract.serializers import AbstractSerializer
from core.user.models import User


class UserSerializer(AbstractSerializer):
    posts_count = serializers.SerializerMethodField()

    def get_posts_count(self, instance):
        return instance.post_set.all().count()

    class Meta:
        model = User
        # List of all the fields that can be included in a request or a response
        fields = ['id', 'username', 'name', 'first_name', 'last_name', 'bio', 'avatar', 'email', 'is_active',
                  'created', 'updated', 'posts_count']
        # List of all the fields that can only be read by the user
        read_only_field = ['is_active']
