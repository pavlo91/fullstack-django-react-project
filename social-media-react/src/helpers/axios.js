import axios from "axios";
import createAuthRefreshInterceptor from "axios-auth-refresh";
import authSlice from "../store/slices/auth";
import store from "../store";

const axiosService = axios.create({
  baseURL: "http://localhost:8000",
  headers: {
    "Content-Type": "application/json",
  },
});

axiosService.interceptors.request.use(async (config) => {
  /**
   * Retrieving the access token from the store and add it to the headers of the request
   */
  const { access } = store.getState().auth;
  config.headers.Authorization = `Bearer ${access}`;
  return config;
});

axiosService.interceptors.response.use(
  (res) => Promise.resolve(res),
  (err) => Promise.reject(err),
);

const refreshAuthLogic = async (failedRequest) => {
  const { refresh } = store.getState().auth;
  return axios
    .post("/refresh/token/", null, {
      baseURL: "http://localhost:8000",
      headers: {
        Authorization: `Bearer ${refresh}`,
      },
    })
    .then((resp) => {
      const { access, refresh } = resp.data;
      failedRequest.response.config.headers["Authorization"] = "Bearer " + access;
      store.dispatch(authSlice.actions.setTokens({ access, refresh }));
    })
    .catch(() => {
      store.dispatch(authSlice.actions.clearTokens());
    });
};

createAuthRefreshInterceptor(axiosService, refreshAuthLogic);

export function fetcher(url) {
  return axiosService.get(url).then((res) => res.data);
}

export default axiosService;
