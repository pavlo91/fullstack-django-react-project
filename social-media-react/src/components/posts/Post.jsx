import React from "react";
import { format } from 'timeago.js';
import {
  LikeFilled,
  CommentOutlined,
  LikeOutlined,
  MoreOutlined,
} from "@ant-design/icons";
import { Image, Card, Dropdown } from "react-bootstrap";
import { RANDOM_AVATAR } from "../../utils";
import { getUser } from "../../hooks/user.actions";

const MoreToggleIcon = React.forwardRef(({ onClick }, ref) => (
  <a
    href="#"
    ref={ref}
    onClick={(e) => {
      e.preventDefault();
      onClick(e);
    }}
  >
    <MoreOutlined />
  </a>
));

function Post(props) {
  const { post } = props;

  const user = getUser();

  return (
    <Card className="rounded-3 my-2">
      <Card.Body>
        <Card.Title className="d-flex flex-row justify-content-between">
          <div className="d-flex flex-row">
            <Image
              src={RANDOM_AVATAR}
              roundedCircle
              width={48}
              height={48}
              className="me-2 border border-primary border-2"
            />
            <div className="d-flex flex-column justify-content-start align-self-center mt-2">
              <p className="fs-6 m-0">{post.author.name}</p>
              <p className="fs-6 fw-lighter">
                <small>{format(post.created)}</small>
              </p>
            </div>
          </div>
          {user.name === post.author.name && (
            <div>
              <Dropdown>
                <Dropdown.Toggle as={MoreToggleIcon}></Dropdown.Toggle>
                <Dropdown.Menu>
                  <Dropdown.Item href="#/action-1">Modifier</Dropdown.Item>
                  <Dropdown.Item href="#/action-2" className="text-danger">
                    Supprimer
                  </Dropdown.Item>
                </Dropdown.Menu>
              </Dropdown>
            </div>
          )}
        </Card.Title>
        <Card.Text>{post.body}</Card.Text>
        <div className="d-flex flex-row">
          <LikeFilled
            style={{
              color: "#fff",
              backgroundColor: "#0D6EFD",
              borderRadius: "50%",
              width: "18px",
              height: "18px",
              fontSize: "75%",
              padding: "2px",
              margin: "3px",
            }}
          />
          <p className="ms-1 fs-6">
            <small>{post.likes_count} like</small>
          </p>
        </div>
      </Card.Body>
      <Card.Footer className="d-flex bg-white w-50 justify-content-between border-0">
        <div className="d-flex flex-row">
          <LikeOutlined
            style={{
              width: "24px",
              height: "24px",
              padding: "2px",
              fontSize: "20px",
              color: post.liked ? "#0D6EFD" : "#C4C4C4",
            }}
          />
          <p className="ms-1">
            <small>Like</small>
          </p>
        </div>
        <div className="d-flex flex-row">
          <CommentOutlined
            style={{
              width: "24px",
              height: "24px",
              padding: "2px",
              fontSize: "20px",
              color: "#C4C4C4",
            }}
          />
          <p className="ms-1">
            <small>Comment</small>
          </p>
        </div>
      </Card.Footer>
    </Card>
  );
}

export default Post;
