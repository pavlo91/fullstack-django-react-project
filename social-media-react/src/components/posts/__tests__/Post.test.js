import { render, screen } from "../../../helpers/test-utils";
import TestRenderer from "react-test-renderer";
import Post from "../Post";
import { setUserData } from "../../../hooks/user.actions";
import userFixtures from "../../../helpers/fixtures/user";
import { BrowserRouter } from "react-router-dom";

const userData = userFixtures();
const postData = {
  id: "ee2a9a31-c4b7-4884-984a-3d55043f8ba2",
  author: {
    id: "0590cd67-eacd-4299-8413-605bd547ea17",
    first_name: "Mossie",
    last_name: "Murphy",
    name: "Mossie Murphy",
    post_count: 3,
    email: "Mossie@yopmail.com",
    bio: "Omnis necessitatibus facere vel in est provident sunt tempora earum accusantium debitis vel est architecto minima quis sint et asperiores.",
    username: "MossieMurphy",
    avatar: null,
    created: "2022-08-19T17:31:03.310Z",
    updated: "2022-08-20T07:38:47.631Z",
  },
  body: "Qui corrupti ut sint provident id et recusandae placeat ad at asperiores odit occaecati praesentium minus id cumque quia molestiae.",
  edited: false,
  liked: true,
  likes_count: 2,
  comments_count: 2,
  created: "2022-08-19T21:10:09.197Z",
  updated: "2022-08-20T05:46:26.549Z",
};

beforeEach(() => {
  // to fully reset the state between __tests__, clear the storage
  localStorage.clear();
  // and reset all mocks
  jest.clearAllMocks();

  setUserData({
    user: userData,
    access: null,
    refresh: null,
  });
});

test("render Post component", () => {
  render(<Post post={postData} />);

  const postElement = screen.getByTestId("post-test");

  expect(postElement).toBeInTheDocument();
});

describe("Post component snapshot", () => {
  it("Matches DOM snapshot", () => {
    const postDomTree = TestRenderer.create(
      <BrowserRouter>
        <Post post={postData} />
      </BrowserRouter>
    ).toJSON();
    expect(postDomTree).toMatchSnapshot();
  });
});
