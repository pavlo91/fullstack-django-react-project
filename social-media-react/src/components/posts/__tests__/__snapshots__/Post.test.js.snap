// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Post component snapshot Matches DOM snapshot 1`] = `
<div
  className="rounded-3 my-4 card"
  data-testid="post-test"
>
  <div
    className="card-body"
  >
    <div
      className="d-flex flex-row justify-content-between card-title h5"
    >
      <div
        className="d-flex flex-row"
      >
        <img
          className="me-2 border border-primary border-2 rounded-circle"
          height={48}
          src={null}
          width={48}
        />
        <div
          className="d-flex flex-column justify-content-start align-self-center mt-2"
        >
          <p
            className="fs-6 m-0"
          >
            Mossie Murphy
          </p>
          <p
            className="fs-6 fw-lighter"
          >
            <small>
              15 hours ago
            </small>
          </p>
        </div>
      </div>
    </div>
    <p
      className="card-text"
    >
      Qui corrupti ut sint provident id et recusandae placeat ad at asperiores odit occaecati praesentium minus id cumque quia molestiae.
    </p>
    <div
      className="d-flex flex-row justify-content-between"
    >
      <div
        className="d-flex flex-row"
      >
        <span
          aria-label="like"
          className="anticon anticon-like"
          role="img"
          style={
            Object {
              "backgroundColor": "#0D6EFD",
              "borderRadius": "50%",
              "color": "#fff",
              "fontSize": "75%",
              "height": "18px",
              "margin": "3px",
              "padding": "2px",
              "width": "18px",
            }
          }
        >
          <svg
            aria-hidden="true"
            data-icon="like"
            fill="currentColor"
            focusable="false"
            height="1em"
            viewBox="64 64 896 896"
            width="1em"
          >
            <path
              d="M885.9 533.7c16.8-22.2 26.1-49.4 26.1-77.7 0-44.9-25.1-87.4-65.5-111.1a67.67 67.67 0 00-34.3-9.3H572.4l6-122.9c1.4-29.7-9.1-57.9-29.5-79.4A106.62 106.62 0 00471 99.9c-52 0-98 35-111.8 85.1l-85.9 311h-.3v428h472.3c9.2 0 18.2-1.8 26.5-5.4 47.6-20.3 78.3-66.8 78.3-118.4 0-12.6-1.8-25-5.4-37 16.8-22.2 26.1-49.4 26.1-77.7 0-12.6-1.8-25-5.4-37 16.8-22.2 26.1-49.4 26.1-77.7-.2-12.6-2-25.1-5.6-37.1zM112 528v364c0 17.7 14.3 32 32 32h65V496h-65c-17.7 0-32 14.3-32 32z"
            />
          </svg>
        </span>
        <p
          className="ms-1 fs-6"
        >
          <small>
            2
             like
          </small>
        </p>
      </div>
      <p
        className="ms-1 fs-6"
      >
        <small>
          <a
            href="/post/ee2a9a31-c4b7-4884-984a-3d55043f8ba2/"
            onClick={[Function]}
          >
            2
             comments
          </a>
        </small>
      </p>
    </div>
  </div>
  <div
    className="d-flex bg-white w-50 justify-content-between border-0 card-footer"
  >
    <div
      className="d-flex flex-row"
    >
      <span
        aria-label="like"
        className="anticon anticon-like"
        onClick={[Function]}
        role="img"
        style={
          Object {
            "color": "#0D6EFD",
            "fontSize": "20px",
            "height": "24px",
            "padding": "2px",
            "width": "24px",
          }
        }
        tabIndex={-1}
      >
        <svg
          aria-hidden="true"
          data-icon="like"
          fill="currentColor"
          focusable="false"
          height="1em"
          viewBox="64 64 896 896"
          width="1em"
        >
          <path
            d="M885.9 533.7c16.8-22.2 26.1-49.4 26.1-77.7 0-44.9-25.1-87.4-65.5-111.1a67.67 67.67 0 00-34.3-9.3H572.4l6-122.9c1.4-29.7-9.1-57.9-29.5-79.4A106.62 106.62 0 00471 99.9c-52 0-98 35-111.8 85.1l-85.9 311H144c-17.7 0-32 14.3-32 32v364c0 17.7 14.3 32 32 32h601.3c9.2 0 18.2-1.8 26.5-5.4 47.6-20.3 78.3-66.8 78.3-118.4 0-12.6-1.8-25-5.4-37 16.8-22.2 26.1-49.4 26.1-77.7 0-12.6-1.8-25-5.4-37 16.8-22.2 26.1-49.4 26.1-77.7-.2-12.6-2-25.1-5.6-37.1zM184 852V568h81v284h-81zm636.4-353l-21.9 19 13.9 25.4a56.2 56.2 0 016.9 27.3c0 16.5-7.2 32.2-19.6 43l-21.9 19 13.9 25.4a56.2 56.2 0 016.9 27.3c0 16.5-7.2 32.2-19.6 43l-21.9 19 13.9 25.4a56.2 56.2 0 016.9 27.3c0 22.4-13.2 42.6-33.6 51.8H329V564.8l99.5-360.5a44.1 44.1 0 0142.2-32.3c7.6 0 15.1 2.2 21.1 6.7 9.9 7.4 15.2 18.6 14.6 30.5l-9.6 198.4h314.4C829 418.5 840 436.9 840 456c0 16.5-7.2 32.1-19.6 43z"
          />
        </svg>
      </span>
      <p
        className="ms-1"
      >
        <small>
          Like
        </small>
      </p>
    </div>
    <div
      className="d-flex flex-row"
    >
      <span
        aria-label="comment"
        className="anticon anticon-comment"
        role="img"
        style={
          Object {
            "color": "#C4C4C4",
            "fontSize": "20px",
            "height": "24px",
            "padding": "2px",
            "width": "24px",
          }
        }
      >
        <svg
          aria-hidden="true"
          data-icon="comment"
          fill="currentColor"
          focusable="false"
          height="1em"
          viewBox="64 64 896 896"
          width="1em"
        >
          <defs>
            <style />
          </defs>
          <path
            d="M573 421c-23.1 0-41 17.9-41 40s17.9 40 41 40c21.1 0 39-17.9 39-40s-17.9-40-39-40zm-280 0c-23.1 0-41 17.9-41 40s17.9 40 41 40c21.1 0 39-17.9 39-40s-17.9-40-39-40z"
          />
          <path
            d="M894 345a343.92 343.92 0 00-189-130v.1c-17.1-19-36.4-36.5-58-52.1-163.7-119-393.5-82.7-513 81-96.3 133-92.2 311.9 6 439l.8 132.6c0 3.2.5 6.4 1.5 9.4a31.95 31.95 0 0040.1 20.9L309 806c33.5 11.9 68.1 18.7 102.5 20.6l-.5.4c89.1 64.9 205.9 84.4 313 49l127.1 41.4c3.2 1 6.5 1.6 9.9 1.6 17.7 0 32-14.3 32-32V753c88.1-119.6 90.4-284.9 1-408zM323 735l-12-5-99 31-1-104-8-9c-84.6-103.2-90.2-251.9-11-361 96.4-132.2 281.2-161.4 413-66 132.2 96.1 161.5 280.6 66 412-80.1 109.9-223.5 150.5-348 102zm505-17l-8 10 1 104-98-33-12 5c-56 20.8-115.7 22.5-171 7l-.2-.1A367.31 367.31 0 00729 676c76.4-105.3 88.8-237.6 44.4-350.4l.6.4c23 16.5 44.1 37.1 62 62 72.6 99.6 68.5 235.2-8 330z"
          />
          <path
            d="M433 421c-23.1 0-41 17.9-41 40s17.9 40 41 40c21.1 0 39-17.9 39-40s-17.9-40-39-40z"
          />
        </svg>
      </span>
      <p
        className="ms-1 mb-0"
      >
        <small>
          Comment
        </small>
      </p>
    </div>
  </div>
</div>
`;
